pgmpy/base/base.py

```diff
@@ -0,0 +1,248 @@
+import inspect
+from abc import ABC, abstractmethod
+from typing import Hashable, Iterable, Optional
+
+import networkx as nx
+
+from pgmpy.base._mixin_roles import _GraphRolesMixin
+
+
+class _CoreGraphABC(ABC):
+    """
+    An abstract class.
+
+    Inheriting from `CoreGraph` automatically includes inheritance from `_CoreGraphABC`.
+    """
+
+    # Constants that the graph class must implement.
+    ALLOWED_EDGES = None
+    IS_DIRECTED = None
+    IS_MULTIGRAPH = None
+
+    def __init_subclass__(cls, **kwargs):
+        super().__init_subclass__(**kwargs)
+
+        if inspect.isabstract(cls):
+            return
+
+        if cls.IS_DIRECTED is None:
+            raise TypeError("IS_DIRECTED must be defined.")
+
+        if cls.IS_MULTIGRAPH is None:
+            raise TypeError("IS_MULTIGRAPH must be defined.")
+
+        if cls.ALLOWED_EDGES is None:
+            raise TypeError("ALLOWED_EDGES must be defined.")
+
+    @abstractmethod
+    def _graph_type_marker(self):
+        """
+        This is a pure virtual method to prevent this class (and CoreGraph) from being instantiated directly.
+        Concrete classes like DAG, PDAG, etc., must implement it.
+        """
+        pass
+
+
+class CoreGraph(_CoreGraphABC, nx.MultiGraph, _GraphRolesMixin):
+    """
+    Base graph class.
+
+    Parameters
+    ----------
+    ebunch : input graph (optional, default: None)
+    latents : set of nodes (default: empty set)
+    exposures : set of nodes (default: empty set)
+    outcomes : set of nodes (default: empty set)
+    roles : dict, optional (default: None)
+
+    Examples
+    --------
+    Create an empty CoreGraph with no nodes and no edges.
+
+    >>> from pgmpy.base import CoreGraph
+    >>> G = CoreGraph()
+
+    Edges and vertices can be passed to the constructor as an edge list.
+
+    >>> [Example_Code]
+
+    **Nodes:**
+
+    Add one node,
+
+    >>> [Example_Code]
+
+    Add a list of nodes,
+
+    >>> [Example_Code]
+
+    **Edges:**
+
+    G can also be grown by adding edges.
+
+    Add one edge,
+
+    >>> [Example_Code]
+
+    Add a list of edges,
+
+    >>> [Example_Code]
+
+    Remove one edge,
+
+    >>> [Example_Code]
+
+    Remove a list of edges,
+
+    >>> [Example_Code]
+
+    **Exposures, Outcomes, and Latents:**
+        [Explain] about difference between Roles and Exposures, Outcomes, and Latents.
+
+    >>> [Example_Code]
+
+    **Roles:**
+
+    >>> [Example_Code]
+
+    """
+
+    # The Constants should be removed once Issue #2383 is merged.
+    # And this Constants must be implemented in each graph model that inherits from `CoreGraph`.
+    ALLOWED_EDGES = {"temp"}
+    IS_DIRECTED = False
+    IS_MULTIGRAPH = True
```

Comment:
I think a simpler way would be to define a single attribute here, say: `SUPPORTED_EDGE_TYPES`. Each inheriting class should define this attribute, and because we don't expect users to use `CoreGraph` directly, we do not need to define attributes like `IS_DIRECTED` or `IS_MULTIGRAPH` because it would be obvious from the object type (i.e., for each of the inheriting classes like `DAG`, `PDAG` we know what kind of graph it would be)